applications/RomApplication/custom_strategies/custom_strategies/modal_derivative_strategy.hpp

```diff
@@ -0,0 +1,1109 @@
+        TSystemVectorType basis;
+        TSparseSpace::Resize(basis, rMassMatrix.size1());
+        const auto current_basis_index = r_model_part.GetProcessInfo()[DERIVATIVE_INDEX];
+
+        // Mass-Orthogonalization using modified Gram-Schmidt method
+        for (std::size_t basis_index = 0; basis_index < current_basis_index; basis_index++){
+
+            // Get previous bases
+            TSparseSpace::SetToZero(basis);
+            this->GetBasis(basis_index, basis);
+
+            // Apply Mass-Orthogonalization w.r.t. previous bases
+            rDx -= (inner_prod(prod(rMassMatrix, basis), rDx) * basis);
+
+        }
+
+        // Mass-Normalization
+        rDx /= std::sqrt(inner_prod(prod(rMassMatrix, rDx), rDx));
+
+        KRATOS_CATCH("")        
+    }
+
+    /**
+     * @brief This function retrieves the basis of given index
+     * @details
+     * { 
+     * } 
+     */  
+    void GetBasis(std::size_t basis_index, TSystemVectorType& rBasis)
+    {
+        KRATOS_TRY
+
+        ModelPart& r_model_part = BaseType::GetModelPart();
+
+        DofsArrayType& r_dof_set = this->pGetBuilderAndSolver()->GetDofSet();
+        bool is_active;
+        
+        for (ModelPart::NodeIterator itNode = r_model_part.NodesBegin(); itNode!= r_model_part.NodesEnd(); itNode++) {
+            ModelPart::NodeType::DofsContainerType& node_dofs = itNode->GetDofs();
+            std::size_t num_node_dofs = node_dofs.size();
+            Matrix& r_rom_basis = itNode->GetValue(ROM_BASIS);
+                
+            // fill the basis vector
+            auto itDof = std::begin(node_dofs);
+            for (std::size_t iDOF = 0; iDOF < num_node_dofs; iDOF++)
+            {
+                is_active = !(r_dof_set.find(**itDof) == r_dof_set.end());
+                if ((*itDof)->IsFree() && is_active)
+                    rBasis((*itDof)->EquationId()) = r_rom_basis(iDOF,basis_index);
+                itDof++;
+            }
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function adds the dynamic derivative constraint
+     * @details
+     * { 
+     * The dynamic derivative LHS is singular since A = K - lambda_i*M,
+     * thus the dynamic derivative constraint has to be applied.
+     * This function adds a Dirichlet constraint on the DOF with the maximum absolute value
+     * } 
+     */  
+    void AddDynamicDerivativeConstraint(TSystemVectorType& rBasis)
+    {
+        KRATOS_TRY
+
+        auto& r_dof_set = this->pGetBuilderAndSolver()->GetDofSet();
+
+        // Find the DOF with the maximum absolute value in the considered basis
+        double max_abs_value = 0.0;
+        for (auto dof_i : r_dof_set)
+        {
+            double temp_abs_value = std::abs(rBasis[dof_i.EquationId()]);
+            if (temp_abs_value > max_abs_value)
+            {
+                max_abs_value = temp_abs_value;
+                mFixedDofIndex = dof_i.EquationId();
+            }
+        }
```

Comment:
I would use the ParallelUtils with reduction here

Reply:
that didn't really work. changed to std::max_element at least